plugins/protocols/xdg-activation.cpp

```diff
@@ -0,0 +1,67 @@
+#include "wayfire/core.hpp"
+#include "wayfire/signal-definitions.hpp"
+#include "wayfire/view.hpp"
+#include <memory>
+#include <wayfire/plugin.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/window-manager.hpp>
+#include "config.h"
+
+extern "C"
+{
+#include <wlr/types/wlr_xdg_activation_v1.h>
+}
+
+class wayfire_xdg_activation_protocol_impl : public wf::plugin_interface_t
+{
+  public:
+    void init() override
+    {
+        xdg_activation = wlr_xdg_activation_v1_create(wf::get_core().display);
+        xdg_activation_request_activate.notify = xdg_activation_handle_request_activate;
+
+        wl_signal_add(&xdg_activation->events.request_activate, &xdg_activation_request_activate);
+    }
+
+    void fini() override
+    {}
+
+    bool is_unloadable() override
+    {
+        return false;
+    }
+
+  private:
+    static void xdg_activation_handle_request_activate(struct wl_listener *listener, void *data)
+    {
+        auto event = static_cast<const struct wlr_xdg_activation_v1_request_activate_event*>(data);
+
+        wayfire_view view = wf::wl_surface_to_wayfire_view(event->surface->resource);
+        if (!view)
+        {
+            LOGI("Could not get view");
+            return;
+        }
+
+        if (!event->token->seat)
+        {
+            LOGI("Denying focus request, seat wasn't supplied");
+            return;
+        }
+
+        if (!view->get_output())
+        {
+            LOGI("View has no output");
+            return;
+        }
+
+        LOGI("Activating view");
+        view->get_output()->focus_view(view, true);
```

Comment:
Focus request is more appropriate here (it will also switch to the target workspace, etc.):

https://github.com/WayfireWM/wayfire/blob/master/src/api/wayfire/window-manager.hpp#L45

With it, you also shouldn't need the check above (about the output)

Reply:
I've changed this, and it still does what it's supposed to.